# Neural Network

## Could we program a computer to learn from this data and then be able to tell the output of any new data it recieves?
## yes, we absolutely can, heres how:
### this shows 5 example sets of 4 inputs with input values per example, and with 1 output per example
### the rule here is that if the first 2 values are the same, and the the 3rd value is opposite to the output of 1 and 2 that the output is 1. if not its 0
### example data:
<img src="set.png" style="width: 600px"/>

### What hapends:
1. set down basics of the Neural Network

2. calculate error rate

3. based on the error rate, automatically make the computer make adustments to the weights

4. do steps 2 and 3 19,999 more times

5. test input

### what to do?
### formula:
### $$ \phi(\sum_{i=1}^n x_i w_i) $$
### $ n $ = ammount of inputs 
### $ x_i $ = certain input, eg imput 1 or 2 or sumthing like that
### $ w_i $ = certain weight eg weight 1 or 2
### $ \phi $ = sigmoid normalizing funtion
### sigmoid normalizing function:
### $$ \phi(x) = \frac{1}{1+e^{-x}} $$

### imports

In [3]:
# imports
import numpy as np
import pandas as pd

### what hapens here:
1. define sigmoid func

2. set up our test arrays

3. set up our dummy weights using a set of random numbers

4. apply the sigmoid function for example 1 and the dummy weights

5. make weigts dynamic

6. calculate functions again

7. calculate error

8. calculate adjustments

9. apply adjustments to weights

10. do steps 6-9 20,000 times



In [5]:
def sigmoid(x):   # this of course is the sigmoid normal function we will use
    return 1 / (1 + np.exp(-x))

table_inputs = np.array([[0,1,0],  # our inputs in a nested array (i guess)
                        [1,1,1],
                        [0,0,1],
                        [1,0,1],
                        [1,1,0],
                        [0,1,1],
                        [0,0,0],
                        [1,0,0]])

table_outputs = np.array([[0,0,0,1,1,1,0,1]]).T  # outputs.  the '.T' particle is for formating purposes

# this is the w_j func
dummy_weights = np.array([
    [-0.62747958],
    [-0.30887855],
    [-0.20646505]])  # this makes the random weights static for the time being so i can use it as an example

print("Dummy weights")
print(dummy_weights)
print(table_inputs[0])
print("output")
print(sigmoid(np.dot(table_inputs[0], dummy_weights)))  # the np.dot funtion is identical to the summaion of the product of the imput and the dummy weights (summated formula from earlier)

Dummy weights
[[-0.62747958]
 [-0.30887855]
 [-0.20646505]]
[0 1 0]
output
[0.4233885]


## What do we do with this?
### Lets Only use the values of example 1 for this. the outputed weight for example 1 is 0.4233885
### Well we now know the values of $ w_1, w_2, w_3 $ and so on, we can use the first formula from earlier to calculate:
### $$ \sum_{i=1}^3 x_i w_i = x_1 w_1 + x_2 w_2 + x_3 w_3$$
### $$ 0(-0.62747958) + 1(-0.308878) + 0(-0.20646505) = -0.30887855 $$
### here are the results on desmos, with the point (-0.30887855, 0.4233885) which if you calculate: $ \phi(-0.30887855) $, it equals 0.4233885
<img src="results.png" />

## What does this mean and what do we do with the info?
### we fist need to find errors, correct the weights, then run it about 20,000 times in order to explain

### using the rate of change (derrivative) of the sigmoid function from earlier,
#### sigmoid derrivative function:
### $$ \phi'(x) = x \times (1-x) $$
### error = output from sample data - $ \phi(x) $
### adjustments = error * input * $ \phi'(outputs) $

In [7]:
def sigmoid_error(x): # derrivative function of sigmoid normal function
    return x * (1 - x)

synaptic_weights = 2 * np.random.random((3,1)) - 1  # randomizing weights

print("weights before training:")
print(synaptic_weights) # comparing weights before training
print(table_inputs)
print("eww")

i=0
while i < 20000:
    outputs = sigmoid(np.dot(table_inputs, synaptic_weights)) # applies the origional formula and gives an output
    error = table_outputs - outputs  # calculates error
    adjustments = error * sigmoid_error(outputs)   # calculates adjustments
    synaptic_weights += np.dot(table_inputs.T, adjustments)  # applies adjustments to weights
    i = i+1

print("weights after")
print(synaptic_weights) # 

weights before training:
[[-0.19630774]
 [ 0.88539977]
 [ 0.19563023]]
[[0 1 0]
 [1 1 1]
 [0 0 1]
 [1 0 1]
 [1 1 0]
 [0 1 1]
 [0 0 0]
 [1 0 0]]
eww
weights after
[[13.19769208]
 [-8.74129516]
 [-8.74129516]]


## Test Imput

In [8]:
test_input = [1,0,1]
print(sigmoid(np.dot(test_input, synaptic_weights)))

[0.98852901]
